standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This specification describes the smart contract that governs RLN memberships, in particular:
+
+- the overall smart contract functionality;
+- contract parameters;
+- the values of those parameters for the initial deployment;
+- the way in which the parameters can be modified.
+
+## Background
+
+Rate-Limiting Nullifiers (RLN) is a ZK-based rate-limiting technique used in Waku.
+Before sending messages, users must register in a smart contract.
+They then attach a ZK proof of valid membership to every message.
+Relaying nodes check message validity and drop invalid messages.
+
+A message is considered valid in terms of RLN if:
+- its sender has an active RLN membership; and
+- the sender hasn't exceeded their rate limit within the current epoch.
+
+RLN is managed by a smart contract that keeps track of active memberships.
+Users interact with the contract to register a new membership and
+to obtain Merkle proofs and the tree root,
+which is necessary for proof generation and verification.
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require uses to put down a deposit.
+The aim of the deposit initially is purely to protect the system from abuse.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+The RLN contract provides the following functionality:
+- register a membership in a given tier;
+- check if a membership is valid;
+- withdraw the deposit for an expired membership;
+- extend an expired membership.
+
+For the `Owner`, the contract provides the following additional functionality:
+- change the modifiable parameters (see parameter table below);
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing;
```

Comment:
Will the term slashing (and how it could be triggered in context of RLN) be obvious to readers?